diff -Naur json-c-0.9.old//configure.in json-c-0.9//configure.in
--- json-c-0.9.old//configure.in	2009-12-15 23:53:18.000000000 +0000
+++ json-c-0.9//configure.in	2009-12-16 00:23:24.000000000 +0000
@@ -1,7 +1,7 @@
 AC_PREREQ(2.52)
 
 # Process this file with autoconf to produce a configure script.
-AC_INIT([json-c], 0.9, [michael@metaparadigm.com])
+AC_INIT([json-c], [0.9], [michael@metaparadigm.com])
 
 AM_INIT_AUTOMAKE(AC_PACKAGE_NAME, AC_PACKAGE_VERSION)
 
